### PR TITLE
ACS-2216: Fix incorrect ContentService Java Interface - should accept content propQName

### PR DESCRIPTION
**Pull request by aymanthefirst**
-  Added  new methods to use propQName instead of PROP_CONTENT to be consistent with the rest of the project. 
-  Also added their corresponding tests.
-  Deprecated methods that use PROP_CONTENT

### CODE REVIEW

File: repository/src/main/java/org/alfresco/repo/content/ContentServiceImpl.java [hunk 1]
```diff
@@ -591,14 +591,24 @@ public boolean isContentDirectUrlEnabled()
      * {@inheritDoc}
      */
     @Override
+    @Deprecated
     public boolean isContentDirectUrlEnabled(NodeRef nodeRef)
+    {
+        return isContentDirectUrlEnabled(nodeRef, ContentModel.PROP_CONTENT);
+    }
```
Comment (PunchedChimera) [inline repository/src/main/java/org/alfresco/repo/content/ContentServiceImpl.java:598]: This should be moved to ContentService.java as a `default` method call.
See `ContentService.requestContentDirectUrl(NodeRef nodeRef, boolean attachment)` for an example.

File: repository/src/main/java/org/alfresco/service/cmr/repository/ContentService.java [hunk 2]
```diff
@@ -170,8 +170,18 @@ public ContentWriter getWriter(NodeRef nodeRef, QName propertyQName, boolean upd
      *
      * @return {@code true} if direct access URLs retrieving is supported for the node, {@code false} otherwise
      */
+    @Deprecated
     boolean isContentDirectUrlEnabled(NodeRef nodeRef);
```
Comment (PunchedChimera) [inline repository/src/main/java/org/alfresco/service/cmr/repository/ContentService.java:174]: As mentioned above. Should be changed to a default method, calling the new implementation, something like this:
```suggestion
    @Deprecated
    default boolean isContentDirectUrlEnabled(NodeRef nodeRef)
    {
        return isContentDirectUrlEnabled(nodeRef, ContentModel.PROP_CONTENT);
    }
```

File: repository/src/main/java/org/alfresco/repo/content/ContentServiceImpl.java [hunk 3]
```diff
@@ -615,14 +625,25 @@ public boolean isContentDirectUrlEnabled(NodeRef nodeRef)
     /**
      * {@inheritDoc}
      */
+    @Override
+    @Deprecated
     public DirectAccessUrl requestContentDirectUrl(NodeRef nodeRef, boolean attachment, Long validFor)
+    {
+        return requestContentDirectUrl(nodeRef, ContentModel.PROP_CONTENT, attachment, validFor);
+    }
```
Comment (PunchedChimera) [inline repository/src/main/java/org/alfresco/repo/content/ContentServiceImpl.java:633]: As above.

File: repository/src/main/java/org/alfresco/service/cmr/repository/ContentService.java [hunk 4]
```diff
@@ -197,8 +223,24 @@ default DirectAccessUrl requestContentDirectUrl(NodeRef nodeRef, boolean attachm
      * @throws UnsupportedOperationException if the store is unable to provide the information.
      */
     @Auditable(parameters = {"nodeRef", "validFor"})
+    @Deprecated
     DirectAccessUrl requestContentDirectUrl(NodeRef nodeRef, boolean attachment, Long validFor);
```
Comment (PunchedChimera) [inline repository/src/main/java/org/alfresco/service/cmr/repository/ContentService.java:227]: Needs a to be a default method (as mentioned above) to call the new method.

File: repository/src/main/java/org/alfresco/service/cmr/repository/ContentService.java [hunk 5]
```diff
@@ -181,11 +191,27 @@ public ContentWriter getWriter(NodeRef nodeRef, QName propertyQName, boolean upd
...
         return requestContentDirectUrl(nodeRef, attachment, null);
     }
 
+    /**
+     * Gets a presigned URL to directly access the content. It is up to the actual store
+     * implementation if it can fulfil this request with an expiry time or not.
+     *
+     * @param nodeRef Node ref for which to obtain the direct access {@code URL}.
+     * @param attachment {@code true} if an attachment URL is requested, {@code false} for an embedded {@code URL}.
+     * @param propertyQName the name of the property, which must be of type <b>content</b>
```
Comment (montgolfiere) [inline repository/src/main/java/org/alfresco/service/cmr/repository/ContentService.java:206]: Trivial: re-order in JavaDoc to match order in method sig

File: repository/src/test/java/org/alfresco/repo/content/ContentServiceImplUnitTest.java [hunk 6]
```diff
@@ -142,6 +145,21 @@ public void testRequestContentDirectUrl_SystemWideIsDisabled()
...
+        try
+        {
+            contentService.requestContentDirectUrl(NODE_REF, QNAME, true, 20L);
+            fail("Expected DirectAccessUrlDisabledException");
+        }
+        catch (DirectAccessUrlDisabledException ex)
+        {
+            verify(mockContentStore, never()).isContentDirectUrlEnabled();
+        }
+    }
```
Comment (PunchedChimera) [inline repository/src/test/java/org/alfresco/repo/content/ContentServiceImplUnitTest.java:161]: Not needed. Just update the method signature in the old test. Once deprecated the old signature should eventually be removed from the product.

File: repository/src/test/java/org/alfresco/repo/content/ContentServiceImplUnitTest.java [hunk 7]
```diff
@@ -153,6 +171,17 @@ public void testRequestContentDirectUrl_SystemWideIsEnabledButStoreIsDisabled()
...
+    public void testRequestContentDirectUrlwithQname_SystemWideIsEnabledButStoreIsDisabled()
+    {
+        setupSystemWideDirectAccessConfig(ENABLED);
+        when(mockContentStore.isContentDirectUrlEnabled()).thenReturn(DISABLED);
+
+        DirectAccessUrl directAccessUrl = contentService.requestContentDirectUrl(NODE_REF, QNAME,true, 20L);
+        assertNull(directAccessUrl);
+        verify(mockContentStore, never()).requestContentDirectUrl(anyString(), eq(true), anyString(), anyString(), anyLong());
+    }
+
```
Comment (PunchedChimera) [inline repository/src/test/java/org/alfresco/repo/content/ContentServiceImplUnitTest.java:184]: As above.

File: repository/src/test/java/org/alfresco/repo/content/ContentServiceImplUnitTest.java [hunk 8]
```diff
@@ -164,6 +193,17 @@ public void testRequestContentDirectUrl_StoreIsEnabledButNotImplemented()
...
+    public void testRequestContentDirectUrlwithQname_StoreIsEnabledButNotImplemented()
+    {
+        setupSystemWideDirectAccessConfig(ENABLED);
+        when(mockContentStore.isContentDirectUrlEnabled()).thenReturn(ENABLED);
+
+        DirectAccessUrl directAccessUrl = contentService.requestContentDirectUrl(NODE_REF, QNAME, true, 20L);
+        assertNull(directAccessUrl);
+        verify(mockContentStore, times(1)).requestContentDirectUrl(anyString(), eq(true), anyString(), anyString(), anyLong());
+    }
+
```
Comment (PunchedChimera) [inline repository/src/test/java/org/alfresco/repo/content/ContentServiceImplUnitTest.java:206]: Not needed. Just update the method signature in the old test. Once deprecated the old signature should eventually be removed from the product.

File: repository/src/test/java/org/alfresco/repo/version/ContentServiceImplTest.java [hunk 9]
```diff
@@ -187,4 +188,48 @@ public void testWhenRequestContentDirectUrlIsNotSupported()
         assertNull(contentService.requestContentDirectUrl(nodeRef, true, null));
         assertNull(contentService.requestContentDirectUrl(nodeRef, true, validFor));
     }
+
+    @Test
+    public void testwithQnameWhenRequestContentDirectUrlIsNotSupported()
```
Comment (PunchedChimera) [inline repository/src/test/java/org/alfresco/repo/version/ContentServiceImplTest.java:193]: As above.

**Review by montgolfiere — CHANGES_REQUESTED** [review]
@aymanthefirst looks good so far. I've added a trivial comment. I haven't reviewed the test methods yet.

One thing missing is to change usages - ie. any existing calls to use new non-deprecated methods (eg. in remote-api).

**Review by PunchedChimera — CHANGES_REQUESTED** [review]
Good first pass. I don't think the extra tests are necessary, they just need to be updated to the new method.
See comments re ContentService.java & ContentServiceImpl.java

File: repository/src/main/java/org/alfresco/repo/content/ContentServiceImpl.java (hunk 1 above)
Comment (mpichura) [reply repository/src/main/java/org/alfresco/repo/content/ContentServiceImpl.java:598]: I wouldn't change interface method in a public project and hardcode that here. 
Very unlikely, but there may be custom implementations for this method which then would be broken. 
Same bit of code resides in `ContentServiceImpl` for this method, which, if you ask me, is safer.

File: repository/src/test/java/org/alfresco/repo/content/ContentServiceImplUnitTest.java (hunk 6 above)
Comment (mpichura) [reply repository/src/test/java/org/alfresco/repo/content/ContentServiceImplUnitTest.java:161]: I would keep the tests - once the deprecated methods are removed - these test of course need to be removed as well.

**Review by skopf — DISMISSED** [review]
Please don't use deprecated methods in the rest api. Instead, the cm:content prop name needs to be hard coded in the rest api, just as we have with the /content endpoint as well for example.
But we should not use these deprecated methods in the rest api.
We want to remove these deprecated methods pretty soon. having code relying on them makes this more difficult.

File: repository/src/main/java/org/alfresco/repo/content/ContentServiceImpl.java (hunk 1 above)
Comment (aymanthefirst) [reply repository/src/main/java/org/alfresco/repo/content/ContentServiceImpl.java]: This should be resolved now but please double check it is what you meant

File: repository/src/main/java/org/alfresco/repo/content/ContentServiceImpl.java (hunk 3 above)
Comment (aymanthefirst) [reply repository/src/main/java/org/alfresco/repo/content/ContentServiceImpl.java]: This should be resolved now but please double check it is what you meant

File: repository/src/main/java/org/alfresco/service/cmr/repository/ContentService.java (hunk 4 above)
Comment (aymanthefirst) [reply repository/src/main/java/org/alfresco/service/cmr/repository/ContentService.java]: This should be resolved now but please double check it is what you meant

File: repository/src/main/java/org/alfresco/service/cmr/repository/ContentService.java (hunk 5 above)
Comment (aymanthefirst) [reply repository/src/main/java/org/alfresco/service/cmr/repository/ContentService.java:209]: Resolved. See https://github.com/Alfresco/alfresco-community-repo/pull/813/commits/ca25e3929b24c9fb3bbb5ff89190930839a09a04

**Review by montgolfiere — CHANGES_REQUESTED** [review]
As per my earlier comment (& Stefan & David )- we need to review & change all usages (tests & remote-api etc) to use new non-deprecated methods.

File: repository/src/test/java/org/alfresco/repo/content/ContentServiceImplUnitTest.java [hunk 10]
```diff
@@ -78,6 +79,8 @@
     private static final String X_AMZ_HEADER_2 = "x-amz-header-2";
     private static final String VALUE_2 = "value2";
 
+    private static final QName QNAME = ContentModel.PROP_CONTENT;
```
Comment (montgolfiere) [inline repository/src/test/java/org/alfresco/repo/content/ContentServiceImplUnitTest.java:82]: Trivial: it's not just any QNAME - maybe rename to PROP_CONTENT_QNAME to be explict.

File: repository/src/test/java/org/alfresco/repo/version/ContentServiceImplTest.java [hunk 11]
```diff
@@ -173,7 +174,7 @@ public void testWhenRequestContentDirectUrlIsNotSupported()
 
         try
         {
-            assertNull(contentService.requestContentDirectUrl(null, true, null));
+            assertNull(contentService.requestContentDirectUrl(null, null, true, null));
```
Comment (montgolfiere) [inline repository/src/test/java/org/alfresco/repo/version/ContentServiceImplTest.java:177]: Add -ve test for what happens if qname is passed in as "null" (and all other args are valid including nodeRef).

**Review by montgolfiere — CHANGES_REQUESTED** [review]
As per my earlier comment (& Stefan & David )- we need to review & change all usages (tests & remote-api etc) to use new non-deprecated methods.

Also, added a couple of other minor suggested changes.

File: repository/src/test/java/org/alfresco/repo/version/ContentServiceImplTest.java (hunk 11 above)
Comment (aymanthefirst) [reply repository/src/test/java/org/alfresco/repo/version/ContentServiceImplTest.java:177]: This should be resolved now but please double check

**Review by PunchedChimera — COMMENTED** [review]
Looks good. Thanks for making the changes I previously requested 👍 
I agree with @montgolfiere, might be worth adding the extra test.
Any tests you're changing/updating could also be refactor to use assertThrows() instead of try-catch.

**Review by KikoUK — APPROVED** [review]

File: repository/src/main/java/org/alfresco/service/cmr/repository/ContentService.java [hunk 12]
```diff
@@ -191,13 +206,46 @@ default DirectAccessUrl requestContentDirectUrl(NodeRef nodeRef, boolean attachm
...
+     * implementation if it can fulfil this request with an expiry time or not.
+     *
+     * @param nodeRef Node ref for which to obtain the direct access {@code URL}.
+     * @param propertyQName the name of the property, which must be of type <b>content</b>
      * @param attachment {@code true} if an attachment URL is requested, {@code false} for an embedded {@code URL}.
      * @param validFor The time at which the direct access {@code URL} will expire.
      * @return A direct access {@code URL} object for the content.
      * @throws UnsupportedOperationException if the store is unable to provide the information.
      */
     @Auditable(parameters = {"nodeRef", "validFor"})
```
Comment (montgolfiere) [inline repository/src/main/java/org/alfresco/service/cmr/repository/ContentService.java:247]: Add "propertQName" to list of @Auditable params ?

File: remote-api/src/main/java/org/alfresco/rest/api/impl/NodesImpl.java [hunk 13]
```diff
@@ -3420,7 +3420,7 @@ public Node unlock(String nodeId, Parameters parameters)
     @Override
     public DirectAccessUrl requestContentDirectUrl(NodeRef nodeRef, boolean attachment, Long validFor)
     {
-        DirectAccessUrl directAccessUrl = contentService.requestContentDirectUrl(nodeRef, attachment, validFor);
+        DirectAccessUrl directAccessUrl = contentService.requestContentDirectUrl(nodeRef, ContentModel.PROP_CONTENT, attachment, validFor);
```
Comment (montgolfiere) [inline remote-api/src/main/java/org/alfresco/rest/api/impl/NodesImpl.java:3423]: This looks good. What about all other usages of the deprecated methods ?

**Review by montgolfiere — CHANGES_REQUESTED** [review]
Please see comment(s) - I assume there are other usages of the deprecated methods ?

File: remote-api/src/main/java/org/alfresco/rest/api/impl/NodesImpl.java (hunk 13 above)
Comment (aymanthefirst) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/NodesImpl.java:3423]: Do you mean I should find methods in other projects such as rest-api-explorer? Because according to IntelliJ, there are no more usages in this project.
image

Comment (montgolfiere) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/NodesImpl.java:3423]: After quick review with @aymanthefirst it looks like all REST APIs go via the NodesImpl after all (including for Deleted Nodes &/or Renditions).

File: repository/src/main/java/org/alfresco/service/cmr/repository/ContentService.java (hunk 12 above)
Comment (aymanthefirst) [reply repository/src/main/java/org/alfresco/service/cmr/repository/ContentService.java]: done

**Comment by montgolfiere** [comment]
@aymanthefirst Thanks for the additional updates LGTM (cc @skopf @PunchedChimera).

**Review by montgolfiere — APPROVED** [review]

File: repository/src/test/java/org/alfresco/repo/content/ContentServiceImplUnitTest.java [hunk 14]
```diff
@@ -133,7 +136,7 @@ public void testRequestContentDirectUrl_SystemWideIsDisabled()
         setupSystemWideDirectAccessConfig(DISABLED);
         try
         {
-            contentService.requestContentDirectUrl(NODE_REF, true, 20L);
+            contentService.requestContentDirectUrl(NODE_REF, PROP_CONTENT_QNAME, true, 20L);
```
[the remark this answers is not in the document]
Comment (aymanthefirst) [reply repository/src/test/java/org/alfresco/repo/content/ContentServiceImplUnitTest.java:138]: Please double-check this is what you meant https://github.com/Alfresco/alfresco-community-repo/pull/813/commits/4df90ca37e9fbafe03eeadcc13376949fd7ef9b5

[the remark this answers is not in the document]
Comment (PunchedChimera) [reply repository/src/test/java/org/alfresco/repo/content/ContentServiceImplUnitTest.java:138]: Yes it is, but you do not need the fail() method in an assertThrows. It will fail by itself if the error isn't thrown.

**Review by PunchedChimera — APPROVED** [review]
Thanks for making those changes
